beacon-chain/state/state-native/setters_validator.go

```diff
@@ -16,42 +17,70 @@ func (b *BeaconState) SetValidators(val []*ethpb.Validator) error {
 	b.lock.Lock()
 	defer b.lock.Unlock()
 
-	b.validators = val
-	b.sharedFieldReferences[types.Validators].MinusRef()
-	b.sharedFieldReferences[types.Validators] = stateutil.NewRef(1)
+	if features.Get().EnableExperimentalState {
+		if b.validatorsMultiValue != nil {
+			b.validatorsMultiValue.Detach(b)
+		}
+		b.validatorsMultiValue = NewMultiValueValidators(val)
+	} else {
+		b.validators = val
+		b.sharedFieldReferences[types.Validators].MinusRef()
+		b.sharedFieldReferences[types.Validators] = stateutil.NewRef(1)
+	}
+
 	b.markFieldAsDirty(types.Validators)
 	b.rebuildTrie[types.Validators] = true
-	b.valMapHandler = stateutil.NewValMapHandler(b.validators)
+	b.valMapHandler = stateutil.NewValMapHandler(val)
 	return nil
 }
 
 // ApplyToEveryValidator applies the provided callback function to each validator in the
 // validator registry.
 func (b *BeaconState) ApplyToEveryValidator(f func(idx int, val *ethpb.Validator) (bool, *ethpb.Validator, error)) error {
-	b.lock.Lock()
-	v := b.validators
-	if ref := b.sharedFieldReferences[types.Validators]; ref.Refs() > 1 {
-		v = b.validatorsReferences()
-		ref.MinusRef()
-		b.sharedFieldReferences[types.Validators] = stateutil.NewRef(1)
-	}
-	b.lock.Unlock()
 	var changedVals []uint64
-	for i, val := range v {
-		changed, newVal, err := f(i, val)
-		if err != nil {
-			return err
+	if features.Get().EnableExperimentalState {
+		b.lock.Lock()
+		defer b.lock.Unlock()
+
+		v := b.validatorsMultiValue.Value(b)
+		for i, val := range v {
+			changed, newVal, err := f(i, val)
```

Comment:
The loop is the same in both paths

Reply:
we do a different thing at the end, inside `if changed`